### ISSUE TITLE
Add option to benchmark.py to run all available benchmarks and store in computer-readable format

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#3382]
Currently, `benchmark.py` can only run one benchmark at a time, and only outputs human-readable text.

It would be convenient to have a `--run-all` option that would run all available benchmarks (or possibly those specified combinatorially by a YAML file?) to make it easy to generate and capture (in a computer-readable manner) comprehensive benchmarks on a new system.

The output could, for example, be a YAML file that makes it easy for both humans to read and to manipulate by converting it into a Python `dict`.

**Comment by peastman** [comment]
If you omit the `--test` argument it runs all the tests except the Amber ones.

**Comment by jchodera** [comment]
But I still have to combinatorially regress over all other arguments?
And there's no machine-readable output, right?

**Comment by peastman** [comment]
Correct.

**Comment by jchodera** [comment]
Would it be OK if I added two command-line arguments?
1. A `--run-all` argument that combinatorially runs all combinations of (platform, test, ensemble, precision)
2. A `--outfile` argument that causes the output to also be captured in a machine- and human-readable format (like YAML)

If we wanted to be more flexible about output formats, we could use [pydantic](https://pydantic-docs.helpmanual.io/), which can support multiple output schemas (e.g. JSON, YAML, etc), but it would introduce another dependency.

I've got access to some hardware for benchmarking, and it would be valuable to be able to capture machine-readable output for our benchmarking purposes.

**Comment by jchodera** [comment]
Alternatively, we may want to pull the benchmark code and data files into the `openmm.app` namespace so that benchmarking can be more easily automated programmatically, or run without needing to find the `$HOME/path/to/anaconda/env/{yourenvname}/share/openmm/examples/` directory, which is a significant challenge for most users. For example, it could be something like
```python
python -m openmm.benchmarks.run
```

**Comment by jchodera** [comment]
I've made some improvements:
* Accept an `--all` argument that combinatorially regresses over all valid combinations of (test, platform, precision, heavy_hydrogens, ensemble)
* Added an `--outfile <filename>` argument that allows the benchmark to be written to a machine-readable JSON or YAML file. We could have this create or append if you want to be able to chain together arbitrary benchmarks.
* Added a `--style <style>` argument to select between a legacy-like output (key: value) and a more readable table:
```
  Test   Platform   Precision   Heavy Hydrogens   dt (fs)   Ensemble   ns/day    
 ─────────────────────────────────────────────────────────────────────────────── 
  gbsa   CUDA       single      False             4.0       NVE        2308.920  
  gbsa   CUDA       mixed       False             4.0       NVE        1997.384  
  gbsa   CUDA       double      False             4.0       NVE        65.046    
  gbsa   CUDA       single      False             4.0       NVT        2198.797  
  gbsa   CUDA       mixed       False             4.0       NVT        1887.733  
  gbsa   CUDA       double      False             4.0       NVT        64.265    
  gbsa   CUDA       single      True              5.0       NVE        1631.845  
  gbsa   CUDA       mixed       True              5.0       NVE        1469.282  
  gbsa   CUDA       double      True              5.0       NVE        79.373    
  gbsa   CUDA       single      True              5.0       NVT        1635.787  
  gbsa   CUDA       mixed       True              5.0       NVT        1460.516  
  gbsa   CUDA       double      True              5.0       NVT        79.527    
  rf     CUDA       single      False             4.0       NVE        1333.140  
  rf     CUDA       mixed       False             4.0       NVE        1178.668  
  rf     CUDA       double      False             4.0       NVE        101.026   
  rf     CUDA       single      False             4.0       NVT        1285.542  
  rf     CUDA       mixed       False             4.0       NVT        1070.895  
  rf     CUDA       double      False             4.0       NVT        100.490   
...
```
Does this sound like it's worth including as a PR? We could punt the refactor into `openmm.benchmarks` to a later update.

**Comment by jchodera** [comment]
I should also extend this to capture information about CPU, GPU, hostname, OpenMM version, and datestamp.

Any idea how to query which GPU we have from Python in a general way?

**Comment by peastman** [comment]
> Any idea how to query which GPU we have from Python in a general way?

Not in advance, but once you create a Context you can query its properties.  CUDA and OpenCL both set a `"DeviceName"` property.  OpenCL also sets a `"OpenCLPlatformName"` property that would be good to record.

**Comment by jchodera** [comment]
Great ideas! Thanks!

This now looks like:
```JSON
{
    "benchmarks": [
        {
            "test": "amber20-dhfr",
            "cutoff": 0.9,
            "constraints": "HBonds",
            "hydrogen_mass": "1.5",
            "ensemble": "NVE",
            "timestep_in_fs": 4.0,
            "precision": "single",
            "platform": "OpenCL",
            "platform_properties": {
                "DeviceIndex": "1",
                "DeviceName": "Intel(R) Iris(TM) Plus Graphics 655",
                "OpenCLPlatformIndex": "0",
                "OpenCLPlatformName": "Apple",
                "Precision": "single",
                "UseCpuPme": "false",
                "DisablePmeStream": "false"
            },
            "steps": 58,
            "elapsed_time": 1.040505,
            "ns_per_day": 19.264491761212103
        }
    ],
    "system": {
        "hostname": "lski2838",
        "timestamp": "2021-12-21T00:20:28.649980",
        "openmm_version": "7.6.0.dev-ad113a0",
        "cpuinfo": "Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz",
        "cpuarch": "i386",
        "system": "Darwin"
    }
}
```

I'm thinking of moving the `system` info within each test record, since this will better position us to run and automatically submit small microbenchmark subsets in the future.

Is the number of CPU threads the platform has elected to use available to query somehow?

**Comment by peastman** [comment]
> Is the number of CPU threads the platform has elected to use available to query somehow?

I don't think that's exposed anywhere.

**Comment by peastman** [comment]
This was implemented in #3887.
